File: src/Python/special_matrix.py
###########################################################
# @Author: HoodUSSEnterprise
# @Date: 2026-06-22 13:20:44
# @LastEditors: HoodUSSEnterprise
# @LastEditTime: 2026-06-22 13:21:31
# @FilePath: \asm_matrix_benchmark\src\Python\special_matrix.py
# @Description:special matrices python code
###########################################################

from typing import List, Union

from .base_matrix import Matrix


###########################################################
# @description: create n×n identity matrix
# @param {int} n
# @return {Matrix}
###########################################################
def identity(n: int) -> "Matrix | None":
    if n <= 0:
        print("Size must be positive")
        return None

    data: List[Union[int, float]] = [0 for _ in range(n * n)]
    for i in range(n):
        data[i * n + i] = 1

    return Matrix(n, n, data)


###########################################################
# @description: create rows×cols zero matrix
# @param {int} rows
# @param {int} cols
# @return {Matrix}
###########################################################
def zeros(rows: int, cols: int) -> "Matrix | None":
    if rows <= 0 or cols <= 0:
        print("Dimensions must be positive")
        return None

    data: List[Union[int, float]] = [0 for _ in range(rows * cols)]
    return Matrix(rows, cols, data)


###########################################################
# @description: create rows×cols ones matrix
# @param {int} rows
# @param {int} cols
# @return {Matrix}
###########################################################
def ones(rows: int, cols: int) -> "Matrix | None":
    if rows <= 0 or cols <= 0:
        print("Dimensions must be positive")
        return None

    data: List[Union[int, float]] = [1 for _ in range(rows * cols)]
    return Matrix(rows, cols, data)

File: src/C++/special_matrix.cpp
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-23 17:39:30
@LastEditors: HoodUSSEnterprise
@LastEditTime: 2026-06-23 19:43:44
@FilePath: \asm_matrix_benchmark\src\C++\special_matrix.cpp
@Description: special matrices c++ code
*************************************************************/
#include "matrix_cpp.h"
#include <cstdlib>
#include <ctime>

/***********************************************************
@description: creates an identity matrix of size n x n.
@param {size_t} n
@return {*}
*************************************************************/
MatrixInt MatrixInt::identity(size_t n)
{
    MatrixInt res(n, n);
    for (size_t i = 0; i < n; i++)
    {
        res.data[i * n + i] = 1;
    }
    return res;
}

/***********************************************************
@description: creates a zero matrix of specified dimensions with all elements set to 0
@param {size_t} r
@param {size_t} c
@return {*}
*************************************************************/
MatrixInt MatrixInt::zeros(size_t r, size_t c)
{
    return MatrixInt(r, c);
}

/***********************************************************
@description: creates a matrix of specified dimensions with all elements set to 1
@param {size_t} r
@param {size_t} c
@return {*}
*************************************************************/
MatrixInt MatrixInt::ones(size_t r, size_t c)
{
    MatrixInt res(r, c);
    for (size_t i = 0; i < r * c; i++)
    {
        res.data[i] = 1;
    }
    return res;
}

/***********************************************************
@description: hat creates a matrix of specified dimensions with random integer elements uniformly distributed in the range [low, high)
@param {size_t} r
@param {size_t} c
@param {int} low
@param {int} high
@return {*}
*************************************************************/
MatrixInt MatrixInt::random(size_t r, size_t c, int low, int high)
{
    static bool seeded = false;
    if (!seeded)
    {
        std::srand(static_cast<unsigned>(std::time(nullptr)));
        seeded = true;
    }

    MatrixInt res(r, c);
    int range = high - low;
    for (size_t i = 0; i < r * c; i++)
    {
        res.data[i] = low + (range > 0 ? std::rand() % range : 0);
    }
    return res;
}
